Reductions use outputs as read-write parameters, so mark them as read as well

naoyam opened the pull request:
I believe this is the right thing to. No alias change in the existing C++ and Python tests.

Commit:
Reductions use outputs as read-write parameters, so mark them as read as
well

Commit:
fix

naoyam commented:
Here's an example with a grouped grid welford. `T13` is declared as an alias of `T4`:

```
    // Alias Allocation - register
    auto& T13 = T4;
    #pragma unroll
    for(nvfuser_index_t i191 = 0; i191 < 8; ++i191) {
      T13[i191] = 0.00000000000000000e+00;
    }
```
`T13` is one of the local welford result tensors:

```
       #pragma unroll
        for(nvfuser_index_t i191 = 0; i191 < 8; ++i191) {
          float T2[1];
          T2[0]
             = __half2float(T1[(i190 * 8) + i191]);
          welfordVectorized<float>(T13[i191], T11[i191], T12[i191], T2[0], T20[0], T18[0]);
        }
```
And they are used in the final grid welford. `T4` as an output, and `T13` as an input. 

```
   T3_reduction.welfordGroup<8, float, int64_t, true>(RefTuple<float, float, float, float, float, float, float, float>(T3[0], T3[1], T3[2], T3[3], T3[4], T3[5], T3[6], T3[7]),
      RefTuple<float, float, float, float, float, float, float, float>(T4[0], T4[1], T4[2], T4[3], T4[4], T4[5], T4[6], T4[7]),
      RefTuple<int64_t, int64_t, int64_t, int64_t, int64_t, int64_t, int64_t, int64_t>(T5[0], T5[1], T5[2], T5[3], T5[4], T5[5], T5[6], T5[7]),
      ConstRefTuple<float, float, float, float, float, float, float, float>(T13[0], T13[1], T13[2], T13[3], T13[4], T13[5], T13[6], T13[7]),
      ConstRefTuple<float, float, float, float, float, float, float, float>(T11[0], T11[1], T11[2], T11[3], T11[4], T11[5], T11[6], T11[7]),
      ConstRefTuple<int64_t, int64_t, int64_t, int64_t, int64_t, int64_t, int64_t, int64_t>(T12[0], T12[1], T12[2], T12[3], T12[4], T12[5], T12[6], T12[7]),
```

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp:
```diff
@@ -666,15 +670,19 @@ class BufferUseDefInfo {
       if (maybe_alloc_info.has_value()) {
+        // Reductions use outputs as read-write parameters, so their
+        // outputs need to be marked as read as well
+        const bool is_read_write = ir_utils::isReductionOp(expr);
```
I think this should be fine when we remove trivial reduction, but for now should we specialize to not include trivial reduce?

naoyam replied:
The alias analysis ignores size-1 (scalar) tensors, so it shouldn't matter.

naoyam replied:
Well, sorry, trivial reductions don't always mean size-1. I'll update the code.

Commit:
another fix

Commit:
Add validation

Commit:
comment

naoyam commented:
Looks like every time I feel lazy and think changes are trivial enough to skip validation, I'm wrong. Fixed position settings and added a validation checker in one of the tests that actually reuses a buffer between an input and output of a grid welford op. It doesn't cause any problem in this case as we implement the grid welford as a block welford followed by a grid welford with intermediate results stored in temporary variables, but that's not the case with the welford + broadcast op.

csarofeen reviewed (approved):
LGTM
